python/src/iceberg/io/boto.py

```diff
@@ -0,0 +1,294 @@
+
+        return self._s3_object.get(Range=range_header)["Body"].read()
+
+    def seek(self, offset: int, whence: Literal[0, 1, 2] = 0) -> None:
+        position_new = offset if whence == 0 else self._position + offset if whence == 1 else self._s3_object.content_length + offset if whence == 2 else None
+
+        if not position_new:
+            raise ValueError(f"Cannot seek to position {offset}, invalid whence: {whence}")
+
+        self._position = position_new
+        return self._position
+
+    def tell(self) -> int:
+        return self._position
+
+    def closed(self) -> bool:
+        return False
+
+    def close(self) -> None:
+        pass
+
+class BotoOutputStream:
+    """A wrapper for writing an S3 Object that abides by the OutputStream protocol
+        
+        Args:
+        s3_object(boto3.resources.factory.s3.Object): An s3 object
+    """
+
+    def __init__(self, s3_object):
+        self._s3_object = s3_object
+
+    def write(self, b: bytes) -> None:
+        """Write to the S3 Object
+        
+        Args:
+            b(bytes): The bytes to write to the S3 Object
+        """
+        self._s3_object.put(Body=b)
+
+    def closed(self) -> bool:
+        """Returns where the stream is closed or not
+
+        Since this is a wrapper for requests to S3, there is no concept of closing, therefore this always returns False
+        """
+        return False
+
+    def close(self) -> None:
+        """Closes the stream
+        
+        Since this is a wrapper for requests to S3, there is no concept of closing, therefore this method does nothing
+        """
+        pass
+
+class BotoInputFile(InputFile):
+    """An input file implementation for the BotoFileIO
+
+    Args:
+        location(str): An S3 URI
+
+    Attributes:
+        location(str): An S3 URI
+
+    Examples:
+        >>> from iceberg.io.boto import BotoInputFile
+        >>> input_file = BotoInputFile("s3://foo/bar.txt")
+        >>> file_content = input_file.open().read()  # Read the contents of the BotoInputFile instance
+    """
+
+    def __init__(self, location: str, session: Session):
+        parsed_location = urlparse(location)  # Create a ParseResult from the uri
+
+        if not parsed_location.scheme.startswith('s3'):
+            raise ValueError(f"Cannot create BotoInputFile, scheme not supported: {parsed_location.scheme}")
+
+        self._bucket = parsed_location.netloc
+        self._path = parsed_location.path.strip("/")
+        self._session = session
+        super().__init__(location=location)
+
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+        file_info = self._file_info()
+        return file_info.size
+
+    def exists(self) -> bool:
+        """Checks whether the location exists"""
+        try:
+            self._session.resource('s3').Bucket(self._bucket).Object(self._path).load()  # raises botocore.exceptions.ClientError with a 404 if it does not exist
```

Comment:
Would this also pull the content? That might be expensive

Reply:
According to this page, it's just a head: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Object.load

but maybe a "head_object()" call would be clearer?  (I too assumed load() would actually load the contents...)

Reply:
Ah you're absolutely right. But since we use the session object directly to create resource clients we'll probably need to add an endpoint argument to the `BotoFileIO` init and then consistently pass it to `resource(...)` which also has an [endpoint_url](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session.resource) argument. Let me update this, thanks @joshuarobinson!